### ISSUE TITLE
etcd v3.3.20 websocket watch API buffering issue

### ISSUE THREAD

**Issue by vitalif** [etcd-io/etcd#11815]
Hi!

There's an issue in etcd 3.3 which makes websocket watches unusable. It seems grpc gateway used in 3.3 does some websocket buffering due to which messages aren't delivered to the client in time.

Sample code (nodejs):
```
const WebSocket = require('ws');

const ws = new WebSocket('ws://127.0.0.1:2379/v3/watch');

ws.on('message', (data) =>
{
    console.log(data);
});

ws.on('open', () =>
{
    ws.send('{"create_request": {"key": "YQ==", "range_end": "Yg=="}}\n');
});
```

With etcd 3.4.7, you immediately receive something like `{"result":{"header":{"cluster_id":"14841639068965178418","member_id":"10276657743932975437","revision":"342","raft_term":"9"},"created":true}}`. With etcd 3.3.20, you get nothing.

In fact, if you reproduce this communication with netcat and a manually formatted websocket frame, you get the response back when netcat closes its socket. The same behaviour is also seen in strace which suggest a buffering problem.

Given that etcd 3.4.7 works fine this issue should be easy to solve. Maybe it is enough to just update the grpc-websocket-proxy...

**Comment by polyrabbit** [comment]
It seems the path to websocket proxy in 3.3 is `/v3beta/`, so you may try `ws://127.0.0.1:2379/v3beta/watch`

https://github.com/etcd-io/etcd/blob/release-3.3/embed/serve.go#L232

I'm also curious how did you format a websocket frame using netcat?

**Comment by vitalif** [comment]
Yeah I've tried v3alpha / v3beta in 3.3, of course. The simplest websocket frame is 0x81 <1 byte length> <4 zero bytes> <message>. I manually formatted it into a file with a text editor and then cat | nc like:

(cat request; sleep 2; cat frame) | nc -w 4  localhost 2379

**Comment by polyrabbit** [comment]
@vitalif Thanks for sharing!

I digged into the code and find the problem lies within package `grpc-gateway` - it pipes response of `client.Watch(ctx)` to a http response writer using `Transfer-Encoding: chunked`, but forgets to separate chunks using a delimiter(`\n`), thus the websocket proxy waits for the chunk infinitely: 

https://github.com/etcd-io/etcd/blob/2f3aa893ecd26841d408dcc699967aeadda5dfaf/vendor/github.com/tmc/grpc-websocket-proxy/wsproxy/websocket_proxy.go#L209-L211

Currently etcd uses `grpc-gateway` v1.3.0, it seems they realized the problem immediately and fixed it in v1.3.1:

https://github.com/grpc-ecosystem/grpc-gateway/commit/e4b8a938efae14de11fd97311e873e989896348c#diff-c90e18d9d959604ec95823fb307c9f65

I updated `grpc-gateway` locally and found the response returned immediately, so here comes the PR (#11843) to fix the issue, @jingyih PTAL

**Comment by socketpair** [comment]
@polyrabbit starting from what etcd version this issue was fixed ?
